packages/@react-stately/menu/src/useSubMenuTriggerState.ts

```diff
@@ -0,0 +1,84 @@
+  /** Key of the trigger item. */
+  triggerKey: Key
+}
+
+export interface SubMenuTriggerState extends OverlayTriggerState {
+  // TODO descriptions
+  // Need some form of isOpen and close for usePopover. Also need closeAll for onClose for menu items, would be kinda weird
+  // to pass in both menuTreeState and the state returned by useSubMenuTriggerState to useMenu.
+  isOpen: boolean,
+  focusStrategy: FocusStrategy | null,
+  open: (focusStrategy?: FocusStrategy | null) => void,
+  close: () => void,
+  closeAll: () => void,
+  // TODO: add setOpen and toggle for type parity with useOverlayTriggerState type that Tray and Popover expect. Perhaps call useOverlayTriggerState in
+  // this hook and just make it controlled? Then spread? Kinda weird since they are unused but needed for type parity.
+  toggle: (focusStrategy?: FocusStrategy | null) => void,
+  setOpen: () => void
+}
+
+// TODO: ideally onClose would be passed in from the Trigger level rather than the menu level so that we could call it
+// TODO descriptions and naming
+// TODO: I'm passing in the parent menu's TreeState now so that we can set expanded keys so a menu can know if it has any sub menus open.
+// The alternative to using expandedKeys would involve making each menu track which "level" it belongs to and have it the look up that index in
+// the MenuTreeState's expandedKeyStack and see if its collection has a matching key. This feels arguably more correct but would then involve
+// having useMenu accept MenuTreeState instead which might not be the worst thing tbh. Open to discussion
+export function UNSTABLE_useSubMenuTriggerState<T>(props: SubMenuTriggerProps, state: MenuTreeState & TreeState<T>): SubMenuTriggerState  {
+  let {triggerKey} = props;
+  let {expandedKeysStack, openSubMenu, closeSubMenu, closeAll, setExpandedKeys, expandedKeys, toggleKey} = state;
+  let [level] = useState(expandedKeysStack?.length + 1);
+  let isOpen = useMemo(() => expandedKeysStack[level - 1] === triggerKey, [expandedKeysStack, triggerKey, level]);
+  let [focusStrategy, setFocusStrategy] = useState<FocusStrategy>(null);
+
+  let open = useCallback((focusStrategy: FocusStrategy = null) => {
+    setFocusStrategy(focusStrategy);
+    setExpandedKeys(new Set([triggerKey]));
+    openSubMenu(triggerKey, level);
+  }, [openSubMenu, level, triggerKey, setExpandedKeys]);
+
+  let close = useCallback(() => {
+    setFocusStrategy(null);
+    if (expandedKeys.has(triggerKey)) {
+      toggleKey(triggerKey);
+    }
+    closeSubMenu(triggerKey, level);
+  }, [closeSubMenu, level, triggerKey, toggleKey, expandedKeys]);
+
+  let toggle = useCallback((focusStrategy: FocusStrategy = null) => {
+    setFocusStrategy(focusStrategy);
+    if (isOpen) {
+      close();
+    } else {
+      open(focusStrategy);
+    }
+  }, [close, open, isOpen]);
+
+  return useMemo(() => ({
+    focusStrategy,
+    isOpen,
+    open,
+    close,
+    closeAll,
+    setOpen: () => {},
+    toggle
```

Comment:
get rid of updating `state.expandedKeys` in favor of passing down level and having each menu use said level to look up if it has a matching triggerKey in the useMenuState expandedKeys stack to determine if it has a open sub menu